source/bluetooth/MicroBitPartialFlashingService.cpp

```diff
@@ -0,0 +1,322 @@
+                    blockNum++;
+                    break;
+                }
+            // blockNum is 7, block is full
+            case 3:
+                {
+                    // Fire write event
+                    MicroBitEvent evt(MICROBIT_ID_PFLASH_NOTIFICATION, FLASH_DATA ,CREATE_AND_FIRE);
+                    // Reset blockNum
+                    blockNum = 0;
+                    break;
+                }
+            default:
+                {
+                    blockNum++;
+                    break;
+                }
+        }
+}
+
+
+/**
+  * Write Event
+  * Used the write data to the flash outside of the BLE ISR
+  */
+void MicroBitPartialFlashingService::partialFlashingEvent(MicroBitEvent e)
+{
+  MicroBitFlash flash;
+
+  switch(e.value){
+    case FLASH_DATA:
+    {
+      /*
+       * Set BLE Mode flag if not already set to boot into BLE mode
+       * upon a failed flash.
+       */
+       MicroBitStorage storage;
+       KeyValuePair* flashIncomplete = storage.get("flashIncomplete");
+       if(flashIncomplete == NULL){
+         uint8_t flashIncomplete = 0x01;
+         storage.put("flashIncomplete", &flashIncomplete, sizeof(flashIncomplete));
+       }
+
+
+      // Flash Pointer
+      uint32_t *flashPointer   = (uint32_t *) ((baseAddress << 16) + offset);
+
+      // If the pointer is on a page boundary erase the page
+      if(!((uint32_t)flashPointer % 0x400))
+          flash.erase_page(flashPointer);
+
+      // Create a pointer to the data block
+      uint32_t *blockPointer;
+      blockPointer = block;
+
+      // Write to flash
+      flash.flash_burn(flashPointer, blockPointer, 16);
+
+      // Update flash control buffer to send next packet
+      uint8_t flashNotificationBuffer[] = {FLASH_DATA, 0xFF};
+      ble.gattServer().notify(partialFlashCharacteristicHandle, (const uint8_t *)flashNotificationBuffer, sizeof(flashNotificationBuffer));
+      break;
+    }
+    case END_OF_TRANSMISSION:
+    {
+      // Write one more packet over the next block: if source embed magic was not previously erased, it will be now!
+      uint32_t *blockPointer;
+      uint32_t *flashPointer   = (uint32_t *) ((baseAddress << 16) + offset + 0x40);
+
+      blockPointer = block;
+      flash.flash_burn(flashPointer, blockPointer, 16);
+
+      // Once the final packet has been written remove the BLE mode flag and reset
+      // the micro:bit
+        MicroBitStorage storage;
+        storage.remove("flashIncomplete");
+        delete &storage;
+        microbit_reset();
+      break;
+    }
+    case MICROBIT_RESET:
+    {
+      MicroBitBLEManager::manager->restartInBLEMode();
+      break;
+    }
+  }
+
+}
+
+const uint8_t  MicroBitPartialFlashingServiceUUID[] = {
+    0xe9,0x7d,0xd9,0x1d,0x25,0x1d,0x47,0x0a,0xa0,0x62,0xfa,0x19,0x22,0xdf,0xa9,0xa8
+};
+
+const uint8_t  MicroBitPartialFlashingServiceCharacteristicUUID[] = {
+    0xe9,0x7d,0x3b,0x10,0x25,0x1d,0x47,0x0a,0xa0,0x62,0xfa,0x19,0x22,0xdf,0xa9,0xa8
+};
```

Comment:
@microbit-sam @martinwork Before we commit these into the annals of time, can I just check these UUIDs are consistent and unique with others we've used? @smartyw I seem to recall you kindly allocated the others for us?

Reply:
UUIDs for custom services don't get registered anywhere so the safest approach is to allocate a fresh UUID using your favourite UUID generator (e.g. uuid on Linux). Consistency of number ranges doesn't matter. More important is to ensure the new service and its UUID is documented but I guess you'll get to that later.

Reply:
Thanks @smartyw. Did we not have some spatial locality for the other UUIDs we used though?

Reply:
"spatial locality". I like that! Yes, we did but I think that's more likely because most were all allocated at more or less the same time. I believe I just let Bluetooth Developer Studio allocate them.... though who knows.... it's now a long time ago. You can either take the risk of colliding with someone else's UUID that they allocated after the micro:bit services were originally put together or avoid that risk altogether with a freshly baked one. They don't have any meaning other than as a unique ID and we do have some that definitely have no "spatial locality" already like the Nordic ones. I vote for a fresh one. If we're voting. Oh, let's have a referendum on the UUID! That's a tried and tested way of making decisions with no possibility of unintended consequences! ;-)

Reply:
The existing microbit ones, except the (Nordic?) UART are of the format E95Dxxxx-251D-470A-A062-FA1922DFA9A8. The PF ones are the same format except they start e97d.